### ISSUE TITLE
2.2.4 - XIAOMI Mijia M365 board - fully functional with EFeru FOC port

### ISSUE THREAD

**Issue by trondin** [RoboDurden/Hoverboard-Firmware-Hack-Gen2.x#97]
Not hoverboard, but anyway - a good  inexpensive board. Few weeks ago was 15 euro and free shipping to Finland.
https://www.aliexpress.com/item/1005007036039813.html
Works very well with EFeru port (UART only), including torque mode.
Geehy APM32F103CBT6. Review is [here](https://mysku.club/blog/misc/100712.html) (use google translate)
The sources come later.

 
![IMG_6711](https://github.com/user-attachments/assets/3d6ac36a-858d-4fec-8bf8-46496d0e3a2f)
![pinout](https://github.com/user-attachments/assets/67b3cbbb-1b68-48b4-8095-1d4ed4ff5c14)
![M365_orig](https://github.com/user-attachments/assets/00730baf-9001-40d5-9468-70b603139c6b)
![IMG_6923](https://github.com/user-attachments/assets/ffa619b6-4557-4e00-a964-dac1c38aa006)

**Comment by RoboDurden** [comment]
Thank you very much for sharing. I guess i should assign the seventh t=target to your board: `target_7=APM32F103`
Then this board would be l=layout 1 Gen2.7.1

As your FOC port is the first single-bldc firmware to support FOC and some Gen2.t.l boards have opamps for two phase currents, it might be worth to make your firmware support multiple-layouts.

Would be nice if your code would support either a working Keil project or PlatformIO project.
The makefile is nice for my online compiler and consumers who do not want to change a lot in code.
But for a more complex rework, it would be better to do it in Keil oder PlatformIO.

**Comment by trondin** [comment]
Project designed with Platformio.
It is easy to do it working with different boards, but I don't nave more boards.
 Problem with EFeru firmware - it works well only with hoverboard motor. 
Should we ask EFeru about way to adjust his algorithm for different motors?

**Comment by I-hate-2FA** [comment]
finally f103 support now? can i easily change the pinout for different board

**Comment by trondin** [comment]
I am not sure, for now I didn't care about that. But is is possible to modify and have hardware definitions in one file.

**Comment by RoboDurden** [comment]
When you publish your source and it can be compiled with PlatformIO, i might have the happiness to test it with a gd32f130 board. But i am still a beginner when it comes to config files for new target MCU...

i drive two of these 1000+Watt bldc nicely with a gen1 board and EFeru FOC: https://www.aliexpress.us/item/3256807568735254.html
And this M365 bldc is 350W just like the hoverboard motors. So i would think that the eferu settings are okay with a wide range of motors.

**Comment by RoboDurden** [comment]
> Should we ask EFeru about way to adjust his algorithm for different motors?

 ,To my knowledge, Emanuel Feru is no longer active with the EFeruFOC firmware, but @Candas1 might be happy to learn that you can port the code to new MCU platforms so quickly.

**Comment by trondin** [comment]
Gate drivers are important, sometimes LIN is inverted. sometimes not. It is only one line in software. But if you have analog driver, very easy go get  magical smoke.

**Comment by Candas1** [comment]
À clone is not another mcu

**Comment by RoboDurden** [comment]
Oh sorry then Candas that i asked for you. If this APM32F103CBT6 is just an STM32F103 clone, then AiLife is correct that we finally can get our STM32F103/GD32F103 Gen2 boards running.

**Comment by Candas1** [comment]
GD32F103 yes, not gd32f130

**Comment by Candas1** [comment]
F103 single motor was done some time ago https://github.com/Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE

**Comment by Candas1** [comment]
You can probably also run this firmware on it https://github.com/Koxx3/SmartESC_STM32_v3

**Comment by Candas1** [comment]
And if you swap the mcu with a gd32f303 you can run this https://github.com/Netzpfuscher/VESC_GD32

**Comment by RoboDurden** [comment]
because VESC is to big for a 32 kB mcu ? :-(
SmartESC... does not look like a FOC firmware to me at first glance.
But with Kambodia and Trondin we now have two FOC firmwares for 32F103 boards ?
My Gen2.x GD32 firmware here by now also has a platform.ini file that works for GD32F130. Maybe that `[env:GD32F130C8]` helps to also compile one of these two firmwares ?

As i have said many times, i am not the guy for these low hardware levels :-/

**Comment by I-hate-2FA** [comment]
actually if you set invert lowside wrong motor will just not move unless you are very unlucky, because the ic have protection

**Comment by trondin** [comment]
Sure, but analog drivers (no IC, transistors only) don't have any protection

**Comment by trondin** [comment]
BTW, code size
![memory](https://github.com/user-attachments/assets/b7cb76d8-79cb-441f-8f6d-65beb91dcaae)

**Comment by trondin** [comment]
code is here https://github.com/trondin/ARM32_FOC_baremetal

**Comment by I-hate-2FA** [comment]
on f103cb size is not an issue, only on mm32 it only have 32k its an issue

**Comment by RoboDurden** [comment]
But 13 kB is great because i can dream that @trondin "somehow" might be able to do GD32F130 port on platformIO. Then all our Gen2.1.x boards with OpAmps for phase currents would get FOC.

@trondin , have you made the stupid test

```
	adc_buffer.rrB = 0;
	adc_buffer.rrC = 0;
```

to check that the phase currents indeed effect a smooth foc spinning of the motor ?

**Comment by RoboDurden** [comment]
I only have a complete Gen2.1.1 hoverboard to spare but shipping from Germany to Finnland would be about 27.3€. Only the two mainboards would cost me 10.5€. 
The 2.1.1 board only has one dual opamp chip, usually one opamp is used for overall currentDC. So i am not sure that this very first gen2 board supports dual phase currents :-/
https://raw.githubusercontent.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/main/Overview/Gen2.1.1.jpg

I also have a complete AT32F403 Gen1 (dual bldc) hoverboard that i can not make any use of at the moment.
Shipping both to Finnland would cost me 28€ :-)
But i am not up do date if by now there is allready an easy io install AT32F403 binary out there. Candas posted so many links that i fear that i am not up to date at all.

And of course i do not want Trondin to do something that he would not be happy to do :-)

**Comment by trondin** [comment]
sorry, I never remember how to use git, now it is public
![pivate](https://github.com/user-attachments/assets/edcf0f7a-bf0a-40f4-aa2c-3fe4382c7317)

**Comment by trondin** [comment]
Shipping cost is terrible, how Chinese shops (like Banggood) manage to send cheap stuff with free shipping from Germany? 
I guess, much more cheaper to order same boards from China. BTW, do you know how to get something from Taobao? What I found costs about 30 euro for delivery plus problem with custom.

**Comment by RoboDurden** [comment]
Thanks, i downloaded your repo and it compiles just fine at first try :-)

```
RAM:   [          ]   3.4% (used 692 bytes from 20480 bytes)
Flash: [=         ]  10.0% (used 13148 bytes from 131072 bytes)
Building .pio\build\genericSTM32F103CB\firmware.bin
```

what were the places that you had to edit the pin configurations apart from defines.h ?
I think the analog channels are hard coded in the EFeru firmware..

I could build a 2.2.1 target_2=GD32F103 test setup and try to make such a board running with your port.
But it only has one dual opamp, so i am not sure that it will support the two phase currents:
https://raw.githubusercontent.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/main/Overview/Gen2.2.1.jpg

And i do not really have the happiness right now to do that for these old gen2.2 boards.

This Gen2.x repo mainly is for offering one firmware for the many different mcu targets.

And you might be able to support MM32SPIN0x target and GenericMM32F103 target in one EFeru port.
So big question for me is: Could you add GD32F130 (and GD32E230) as well.
I think the problem is the missing hal library for these targets ?

**Comment by RoboDurden** [comment]
Thanks, i downloaded your repo and it compiles just fine at first try :-)

```
RAM:   [          ]   3.4% (used 692 bytes from 20480 bytes)
Flash: [=         ]  10.0% (used 13148 bytes from 131072 bytes)
Building .pio\build\genericSTM32F103CB\firmware.bin
```

what were the places that you had to edit the pin configurations apart from defines.h ?
I think the analog channels are hard coded in the EFeru firmware..

I could build a 2.2.1 target_2=GD32F103 test setup and try to make such a board running with your port.
![PXL_20240927_092709740_copy_3024x2268 1](https://github.com/user-attachments/assets/c5c371e2-f050-4831-9738-6ac305c5b0da)
But it only has one dual opamp, so i am not sure that it will support the two phase currents:
https://raw.githubusercontent.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/main/Overview/Gen2.2.1.jpg

And i do not really have the happiness right now to do that for these old gen2.2 boards.

This Gen2.x repo mainly is for offering one firmware for the many different mcu targets.

And you might be able to support MM32SPIN0x target and GenericMM32F103 target in one EFeru port.
So big question for me is: Could you add GD32F130 (and GD32E230) as well.
I think the problem is the missing hal library for these targets ?

**Comment by I-hate-2FA** [comment]
single opamp  will be for 2 phase
because if you dont do 100% you can calculate 3 phase from 2
and the currentDC can be read without a opamp

**Comment by trondin** [comment]
HAL is not problem, because I do not use HAL :)
For STM32F103 only a system clock initialization and delay are from HALL.
More important things are bare metal.
MM32SPIN05 port uses registers only.
And projects are really very different because of hardware. MM32SPIN05 ADC is more comfortable for me and no needs to use DMA.
For STM32F103 I used "dummy" timer and DMA because of STM32 ADC limitation,
It is not so easy to combine both project and, I think, it is useless.
Basically my template can be used for any processor with like STM32 advanced timer, but you have to learn hardware carefully.

**Comment by RoboDurden** [comment]
I guess i should assign Gen2.2.4 to this board - if this is indeed a F103 clone.
But on your photo i only see

![grafik](https://github.com/user-attachments/assets/d5d9e53d-0db1-4059-b361-3288a7553ea5)

How can people identify this as a F103 mcu ?

**Comment by trondin** [comment]
Geehy APM32F103CBT6 is a full clone of STM32F103CBT6. You will see difference only during uploading with platformio.
Should be
`upload_flags = -c set CPUTAPID 0x2ba01477`
in ini file.

**Comment by RoboDurden** [comment]
The board identifier here asks the user to identify the mcu of their board: https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/wiki
But i do not see a F103 on your mcu :-/
How did you find out this ARM32FEPSC8T6 is actually a F103 clone ??

**Comment by I-hate-2FA** [comment]
STM is also like this
it is marked STM32FEBKc6t6

**Comment by trondin** [comment]
Honestly I don't remember.
And ARM32FEPSC8T6 do not exist at all, but it is APM32F103CBT6, I don't know what is wrong with marking.
![mcu](https://github.com/user-attachments/assets/7c7d7f2b-cb68-41fc-af66-fcf903ad79c6)

**Comment by I-hate-2FA** [comment]
it does not matter, the internal all same

**Comment by noeyvert** [comment]
I can't even run a motor as a test or receive feedback data in APM32FEPSCBT6

Hello. Congratulations Trondin on your project.
Could you please help me?

I can't even run a motor as a test or receive feedback data.

My board is Geehy APM32FEPSCBT6 as in the image 

<img width="1442" height="764" alt="Image" src="https://github.com/user-attachments/assets/4f723978-bf7b-4401-b43d-93a0434da391" />

**Comment by noeyvert** [comment]
I even edited main.c to:
- force the activation of the motors

<img width="1227" height="245" alt="Image" src="https://github.com/user-attachments/assets/9044a7af-3c98-44b4-b55c-f1f510f57855" />


- force the motor to rotate

<img width="1031" height="256" alt="Image" src="https://github.com/user-attachments/assets/65182917-341c-488d-b05f-08f74abbe653" />

- and send feedback without criteria:

<img width="1208" height="434" alt="Image" src="https://github.com/user-attachments/assets/da587e48-3757-4d83-b07d-6f32d9c6895b" />
